### PR TITLE
RAD2RAD compatibility with NLOCAL

### PR DESCRIPTION
**Pull request by bertrand-altair**
Compatibility between rad2rad and NLOCAL method. NLOCAL dof must be coupled using rad2rad, in the same way than classical dof. New exchanges between rad2rad and engine is needed.

### CODE REVIEW

File: engine/source/coupling/rad2rad/r2r_init.F [hunk 1]
```diff
@@ -236,8 +247,34 @@ SUBROUTINE R2R_INIT(IEXLNK  ,ITAB,IGRNOD,X      ,
...
+          POSI => NLOC_DMG%POSI(1:NLOC_DMG%NNOD)
+          COMPT = 0
+          MY_ALLOCATE(NDOF_NL,NNG)
+          DO I=1,NNG
+            NN = IDXI(GRNOD(I))
+            NDOF_NL(I) = POSI(NN+1)-POSI(NN)
+            COMPT = COMPT + NDOF_NL(I)
+          ENDDO
+          NBDOF_NL(IEX) = COMPT
+          ALLOCATE(IADD_NL(NBDOF_NL(IEX)))
```
Comment (sebastienVilleneuve) [inline engine/source/coupling/rad2rad/r2r_init.F:262]: @bertrand-altair

 IADD_NL is allocated here but never deallocated

File: starter/source/coupling/rad2rad/r2r_group.F [hunk 2]
```diff
@@ -417,17 +419,42 @@ SUBROUTINE R2R_GROUP(
...
+C-----------Nodes in link type 4 + nlocal dof -> aditional link------C
+C-----------Node coupled only if nl material on both sides ----------C 
+              DO J=1,NUMNOD
+                IF ((TAG_NLOCAL(J)==1).AND.(TAGNO(J+NPART+NUMNOD) == N+1)) THEN
+                  BUF_NOD(IAD)=J
+                  IAD=IAD+1
+                  COMPT = COMPT+1	        
+                ENDIF
+              ENDDO
+              DEALLOCATE(TAG_NLOCAL)
```
Comment (sebastienVilleneuve) [inline starter/source/coupling/rad2rad/r2r_group.F:455]: @bertrand-altair 

I'm not sure dealloc it at good position here  : if K < 5 , it is never deallocated

File: starter/source/materials/fail/nloc_dmg_init.F [hunk 3]
```diff
@@ -244,6 +244,14 @@ SUBROUTINE NLOC_DMG_INIT(ELBUF_TAB,NLOC_DMG ,IPARG    ,IXC      ,
...
         ALLOCATE( INDEX(NUMELS+NUMELC+NUMELTG) )
         ALLOCATE( ITRI(NUMELS+NUMELC+NUMELTG) )
+C
+        IF (NSUBDOM > 0) THEN
+C--       multidomains - origianl nummat is used
+          MATSIZE = NUMMAT0
+        ELSE
+          MATSIZE = NUMMAT
+        ENDIF
+        MY_ALLOCATE2D(WARN_LENGHT,MATSIZE,3)
```
Comment (sebastienVilleneuve) [inline starter/source/materials/fail/nloc_dmg_init.F:254]: @bertrand-altair 

WARN_LENGHT is allocated here , but never deallocated

**Review by sebastienVilleneuve — CHANGES_REQUESTED** [review]
@bertrand-altair 

I think I have found some other alloc/dealloc issues
I let you check please .

Thanks,
Sébastien

File: engine/source/coupling/rad2rad/r2r_init.F (hunk 1 above)
Comment (bertrand-altair) [reply engine/source/coupling/rad2rad/r2r_init.F]: fixed

File: starter/source/coupling/rad2rad/r2r_group.F (hunk 2 above)
Comment (bertrand-altair) [reply starter/source/coupling/rad2rad/r2r_group.F]: Not a bug as when array is allocated K=5 is always reached, but i have clean the code anyway.

File: starter/source/materials/fail/nloc_dmg_init.F (hunk 3 above)
Comment (bertrand-altair) [reply starter/source/materials/fail/nloc_dmg_init.F:254]: fixed

**Review by sebastienVilleneuve — APPROVED** [review]

**Pull request merged**